src/main/java/com/cupid/jikting/member/service/CustomOAuth2UserService.java

```diff
@@ -0,0 +1,74 @@
+import com.cupid.jikting.member.oauth2.CustomOAuth2User;
+import com.cupid.jikting.member.oauth2.OAuthAttributes;
+import com.cupid.jikting.member.repository.MemberRepository;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.security.core.authority.SimpleGrantedAuthority;
+import org.springframework.security.oauth2.client.userinfo.DefaultOAuth2UserService;
+import org.springframework.security.oauth2.client.userinfo.OAuth2UserRequest;
+import org.springframework.security.oauth2.client.userinfo.OAuth2UserService;
+import org.springframework.security.oauth2.core.OAuth2AuthenticationException;
+import org.springframework.security.oauth2.core.user.OAuth2User;
+import org.springframework.stereotype.Service;
+
+import java.util.Collections;
+import java.util.Map;
+
+@Slf4j
+@RequiredArgsConstructor
+@Service
+public class CustomOAuth2UserService implements OAuth2UserService<OAuth2UserRequest, OAuth2User> {
+
+    private final MemberRepository memberRepository;
+
+    private static final String KAKAO = "kakao";
+
+    @Override
+    public OAuth2User loadUser(OAuth2UserRequest userRequest) throws OAuth2AuthenticationException {
+        log.info("CustomOAuth2UserService.loadUser() 실행 - OAuth2 로그인 요청 진입");
+        OAuth2UserService<OAuth2UserRequest, OAuth2User> delegate = new DefaultOAuth2UserService();
+        OAuth2User oAuth2User = delegate.loadUser(userRequest);
+        String registrationId = userRequest.getClientRegistration().getRegistrationId();
+        SocialType socialType = getSocialType(registrationId);
+        String userNameAttributeName = userRequest.getClientRegistration()
+                .getProviderDetails()
+                .getUserInfoEndpoint()
+                .getUserNameAttributeName();
+        Map<String, Object> attributes = oAuth2User.getAttributes();
+        OAuthAttributes extractAttributes = OAuthAttributes.of(socialType, userNameAttributeName, attributes);
+        Member createdMember = getMember(extractAttributes, socialType);
+        return new CustomOAuth2User(
+                Collections.singleton(new SimpleGrantedAuthority(createdMember.getRole().getKey())),
+                attributes,
+                extractAttributes.getNameAttributeKey(),
+                createdMember.getUsername(),
+                createdMember.getRole()
+        );
+    }
+
+    private SocialType getSocialType(String registrationId) {
+        if (KAKAO.equals(registrationId)) {
+            return SocialType.KAKAO;
+        }
+        return SocialType.GOOGLE;
+    }
+
+    private Member getMember(OAuthAttributes attributes, SocialType socialType) {
+        Member findMember = memberRepository.findBySocialTypeAndSocialId(socialType, attributes.getOauth2UserInfo().getId())
+                .orElse(null);
+        if (findMember == null) {
+            return saveMember(attributes, socialType);
+        }
+        return findMember;
+    }
```

Comment:
p2. Optional의 orElseGet을 아래처럼 사용하는게 맞는지 정확히 기억이 안나는데, 이런 식으로 코드를 줄일 수 있지 않을까요?

```suggestion
    private Member getMember(OAuthAttributes attributes, SocialType socialType) {
        return memberRepository.findBySocialTypeAndSocialId(socialType, attributes.getOauth2UserInfo().getId())
                .orElseGet(() -> saveMember(attributes, socialType));
    }
```

Reply:
수정완료했습니다